proposals/dark_mode_extension_icons.md

```diff
@@ -0,0 +1,231 @@
+
+  Optionally the `color_scheme` matching criterium which can either be an string or
+array of color schemes.
+
+Setting `icon_variants` dynamically:
+```
+const exampleProperties: {string: string | ImageData}[] = [
+  {
+    "any": "any.svg",
+  },
+  {
+    "16": "16.png",
+    "32": "32.png"
+  },
+  {
+    "16": darkImageData16,
+    "32": darkImageData32,
+    "color_scheme": "dark"
+  },
+  {
+    "16": lightImageData16,
+    "32": lightImageData32,
+    "color_scheme": "light"
+  }
+];
+
+// action.setIcon().
+const createProperties = {variants: exampleProperties};
+action.setIcon(createProperties);
+
+// menus.*(), for supporting browsers.
+const menusProperties = {icon_variants: exampleProperties};
+menus.create(menusProperties);
+menus.update(menusProperties);
+```
+
+A benefit of this new structure is that it's more resiliant to future changes,
+thus allowing for more keys such as density (e.g. 2dppx), purpose (e.g.
+monochrome), and etc.
+
+## setIcon()
+Incumbent action.setIcon(), for reference.
+```
+action.setIcon({
+  path,
+  tabId
+  imageData,
+});
+```
+
+### Behavior
+
+Existing manifest key. The behavior of the icons manifest key will remain
+unchanged.
+```
+{
+  "icons": {
+    "64": "icon_64.png"
+  },
+  "action": {
+    "default_icon": {
+      "64": "action_64.png"
+    }
+  }
+}
+```
+
+Order of precedence. The new `icon_variants` keys and subkeys will take
+precedent, followed by the incumbent `icons` key.
+
+The `dark` icon variant can be used by browsers to improve readability of the
+icon. Often this will be when the OS or browser color scheme is dark (often
+referred to as "dark mode") but also when the browser uses a darker shade theme
+which has a background on which the dark icon variant would be more readable.
+The `light` icon variant is used otherwise.
+
+### New Permissions
+
+N/A.
+
+### Manifest File Changes
+
+1. Any icon group that does not contain a "color_scheme" key will apply to all
+available options, e.g. both "light" and "dark".
+
+1. If `icon_variants` contains an icon group with matching conditions, the icon(s)
+specified in the first matching icon group based on insertion order will be used.
+The other icon groups after that match will be ignored.
+   * If there is more than one matching icon object, any without `color_scheme` will
+be applied to every possible color scheme. Therefore, a subsequent matching icon
+object with a `color_scheme` will not be used.
+   * Incorrectly typed `color_scheme` values will be ignored, with an optional warning. For example, `color_scheme: ["unknown"]` will be treated as an empty array.
+   * If only one icon object is defined with a specific color scheme, that
+   icon object will be applied to all color schemes. It will be the icon used.
+
+1. If only one icon group is supplied, all matching conditions will be ignored.
+1. icon_variants will not cause an error in the event that it is invalid
```

Comment:
Although this section is about the manifest, it also mentions setIcon.

Therefore I am asking the question here: can we be explicit about what happens when an unrecognized property (or unrecognized color scheme) is found in the API?

Chrome and Firefox have the convention in the extension API to throw from API methods, which enables feature detection at runtime.

Safari might be inclined to silently ignore unknown keys.

@xeenon thoughts?

Reply:
`color_scheme` would be a known key, so we would throw for unknown values. We _would not_ throw for other keys (other than numbers or `color_scheme`).

Reply:
@xeenon correct me if I am wrong, but I believe you mentioned unknown values in `color_scheme` should just be ignored by the browser? Why should an error be thrown in this situation?

Reply:
@carlosjeurissen It would be ignored in the manifest, but I support it in the JS API, since that allows feature detection in JS.

Reply:
@xeenon It would be ideal if we can accommodate both the feature detection and the main benefit of this new syntax being the forgiving / future proof syntax. I believe we can achieve this by going for the following:

There are two kind of matching criteria. One being the property and one being values. When passing matching criteria property names in icon groups to `action.setIcon` which are not understood by the browser, the whole icon group will be skipped and marked as not matching. For property values, for properties taking multiple values (like `color_scheme`, the property values not understood by the browsers will silently be ignored. A property which ends up with no supported values will also be considered unsupported.

So this would mean `action.setIcon` would not throw an exception at all. Except, and here is where the feature detection comes in. If non of the icon groups have supported matching criteria. Instead of dropping all matching criteria as would be done for the manifest (I believe we agreed on this? might need some more specification), when no icon groups have supported matches, `action.setIcon` will throw an exception. 

This behaviour would mean feature detection can be done by passing one icon group. For example:
```js
browser.action.setIcon({
  variants: [{
    any: "path.svg",
    color_scheme: "dark"
  }]
}).catch((e) => {
  // throws if the browser does not support dark color scheme since no icon group is supported.
});
```

And still use the forgiving fallback nature of the syntax as such:
```js
browser.action.setIcon({
  variants: [{
    any: "path.svg",
    color_scheme: "dark"
  }, {
    16: "path.png"
}]
}).then(result => {
  // even tho the browser does not support dark color_scheme, it will not throw and simply ignore this icon group. 
});
````

This way we have the best of both worlds.

Reply:
@carlosjeurissen I like that! Makes sense to me.

Reply:
A related "commit suggestion" was applied to the patch, which addresses this subthread:
```
`action.setIcon({variants})` will not throw when giving it icon groups with
properties it does not understand. Those icon groups will simply be ignored for
making matches. If however, none of the icon groups are supported, an exception
will be thrown allowing both feature detection and specifying fallbacks without
requiring feature detection.
```